File: node/src/lazy.js
function* generator(n) {
    while (true) {
        yield n
        n++
    }
}

function findFirst(iterator, predicate) {
    for (const i of iterator) {
        if (predicate(i)) {
            return i
        }
    }
}

const iterator = generator(0)
const element = findFirst(iterator, n => n >= 2)
console.log(element) // 2

File: java/src/Lazy.java
import java.util.stream.Stream;

class Lazy {
    public static void main(String[] args) {
        var iterator = Stream.iterate(0, n -> n + 1);
        var element = iterator
                .filter(n -> n >= 2)
                .findFirst()
                .orElse(null);
        System.out.println(element); // 2
    }
}
